File: src/icon/serialization/deserializer.py
import logging
from typing import TYPE_CHECKING, Any, cast

import pydantic
import pydase
import pydase.components
import pydase.utils.serialization.deserializer
from pydase.utils.serialization.types import SerializedObject

from icon.serialization.types import (
    SerializedIconObject,
    SerializedPydanticModel,
)

if TYPE_CHECKING:
    from collections.abc import Callable

logger = logging.getLogger(__name__)


class IconDeserializer(pydase.utils.serialization.deserializer.Deserializer):
    """
    This deserializer adds deserialization of pydantic models to the
    `pydase.utils.serialization.deserializer.Deserializer`.
    """

    @classmethod
    def deserialize(cls, serialized_object: SerializedIconObject) -> Any:
        type_handler: dict[str | None, None | Callable[..., Any]] = {
            None: None,
            "int": cls.deserialize_primitive,
            "float": cls.deserialize_primitive,
            "bool": cls.deserialize_primitive,
            "str": cls.deserialize_primitive,
            "NoneType": cls.deserialize_primitive,
            "Quantity": cls.deserialize_quantity,
            "Enum": cls.deserialize_enum,
            "ColouredEnum": lambda serialized_object: cls.deserialize_enum(
                serialized_object, enum_class=pydase.components.ColouredEnum
            ),
            "list": cls.deserialize_list,
            "dict": cls.deserialize_dict,
            "method": cls.deserialize_method,
            "Exception": cls.deserialize_exception,
            "datetime": cls.deserialize_datetime,
            "pydantic.BaseModel": cls.deserialize_pydantic_basemodel,
        }

        # First go through handled types (as ColouredEnum is also within the components)
        handler = type_handler.get(serialized_object["type"])
        if handler:
            return handler(serialized_object)

        # Custom types like Components or DataService classes
        service_base_class = cls.get_service_base_class(serialized_object["type"])
        if service_base_class:
            return cls.deserialize_data_service(
                cast(SerializedObject, serialized_object), service_base_class
            )

        return None

    @classmethod
    def deserialize_pydantic_basemodel(
        cls, serialized_object: SerializedPydanticModel
    ) -> pydantic.BaseModel:
        def get_model_from_module(module: str, model_name: str) -> pydantic.BaseModel:
            import importlib

            return getattr(importlib.import_module(module), model_name)

        class_module, class_name = serialized_object["name"].rsplit(".", 1)
        return get_model_from_module(class_module, class_name).model_validate_json(
            serialized_object["value"]
        )


def loads(serialized_object: SerializedIconObject) -> Any:
    return IconDeserializer.deserialize(serialized_object)

File: frontend/src/utils/deserializer.ts
import { SignJsonWebKeyInput } from "crypto";
import {
  SerializedDataService,
  SerializedDict,
  SerializedEnum,
  SerializedException,
  SerializedList,
  SerializedObject,
} from "../types/SerializedObject";

type DeserializedValue =
  any; /* eslint-disable-line @typescript-eslint/no-explicit-any */

class Deserializer {
  private static typeHandler: Record<
    SerializedObject["type"],
    | ((
        serializedObject: any /* eslint-disable-line @typescript-eslint/no-explicit-any */,
      ) => DeserializedValue)
    | undefined
  > = {
    int: Deserializer.deserializePrimitive,
    float: Deserializer.deserializePrimitive,
    bool: Deserializer.deserializePrimitive,
    str: Deserializer.deserializePrimitive,
    NoneType: Deserializer.deserializePrimitive,
    Enum: Deserializer.deserializeEnum,
    ColouredEnum: Deserializer.deserializeEnum,
    list: Deserializer.deserializeList,
    dict: Deserializer.deserializeDict,
    None: Deserializer.deserializePrimitive,
    DataService: Deserializer.deserializeDataService,
    Image: Deserializer.deserializeDataService,
    NumberSlider: Deserializer.deserializeDataService,
    DeviceConnection: Deserializer.deserializeDataService,
    Task: Deserializer.deserializeDataService,
    Exception: Deserializer.deserializeException,
    Quantity: undefined,
  };

  static deserialize(serializedObject: SerializedObject): DeserializedValue {
    const handler = this.typeHandler[serializedObject.type];
    if (handler) {
      return handler(serializedObject);
    }

    console.warn(`Unknown type: ${serializedObject.type}`);
    return null;
  }

  private static deserializePrimitive(
    serializedObject: SerializedObject,
  ): DeserializedValue {
    if (serializedObject.type === "float") {
      return parseFloat(serializedObject.value as unknown as string);
    }
    return serializedObject.value;
  }

  private static deserializeException(serializedObject: SerializedException): Error {
    const error = new Error(serializedObject.value);
    error.name = serializedObject.name;
    return error;
  }

  private static deserializeEnum(serializedObject: SerializedEnum): DeserializedValue {
    const { enum: enumValues, value } = serializedObject;
    return enumValues[value];
  }

  private static deserializeList(serializedObject: SerializedList): DeserializedValue {
    return serializedObject.value.map((item) => Deserializer.deserialize(item));
  }

  private static deserializeDict(serializedObject: SerializedDict): DeserializedValue {
    const deserialized: Record<string, DeserializedValue> = {};
    for (const [key, value] of Object.entries(serializedObject.value)) {
      deserialized[key] = Deserializer.deserialize(value);
    }
    return deserialized;
  }

  private static deserializeDataService(
    serializedObject: SerializedDataService,
  ): DeserializedValue {
    const proxy: Record<string, DeserializedValue> = {};
    for (const [key, value] of Object.entries(serializedObject.value)) {
      proxy[key] = Deserializer.deserialize(value);
    }
    return proxy;
  }
}

// Usage
export function deserialize(serializedObject: SerializedObject): DeserializedValue {
  return Deserializer.deserialize(serializedObject);
}
